forem/forem
# Allow sign up without email smtp config

## What type of PR is this? (check all applicable)

## Description
For FOSS users. This mimics a development-like environment where we allow the app to function without SMTP configuration. This whole thing feels like a hack so I want to point a few crucial points.

- I had to modify the creator sign-up flow. Instead of `sign up -> confirmation page -> click email confirmation -> open browser and log in`, it's now just two steps for FOSS, `signup -> logged-in and onboarding`. Since we can't send an email, the user will be routed directly to the `sign-in` page. I have no made any style or alert change so the transition could be a bit jarring.

- I also have not provided a way for FOSS admin to enter their own SMTP credential via the `/admin` since that is not the scope of this PR.

- When a user signed up via Github or Twitter, we confirm their email by using `user.skip_confirmation!`. My changes here keep them unconfirmed (`confirmed_at = nil`). This could be the first time this kind of implementation is introduced to this app and I don't think it will create any problem. See [Devise's wiki](https://github.com/heartcombo/devise/wiki/How-To:-Add-:confirmable-to-Users#allowing-unconfirmed-access) for more info.

## Related Tickets & Documents
n/a

## QA Instructions, Screenshots, Recordings
You will need to mimic the Creator environment. For the best result, you will need to reset your database in step 4

1. Add `export MODE="STARTER"` to your `.env`.
2. Add `export FOREM_OWNER_SECRET="test"` to your `.env`, you'll need this during sign-up process.
3. Enable `raise_delivery_errors` in your development.rb ie `config.action_mailer.raise_delivery_errors = true`
4. Reset your database and set it up again `bin/rails db:reset && bin/setup`.
5. Startup the server and navigate to the `localhost:3000`.
6. Try signing up, you should be redirected to the sign-in page immediately and able to log in without issue.
 
## Added tests?
- [x] Updated existed spec

## [Forem core team only] How will this change be communicated?
- [ ] I need to update our FOSS documentation.

## [optional] Are there any post deployment tasks we need to perform?
n/a
